src/Markdig/Helpers/CharHelper.cs

```diff
@@ -306,5 +306,376 @@ public static bool IsEmailUsernameSpecialChar(char c)
         {
             return ".!#$%&'*+/=?^_`{|}~-+.~".IndexOf(c) >= 0;
         }
+
+        [MethodImpl(MethodImplOptionPortable.AggressiveInlining)]
+        public static bool IsRightToLeft(char c)
+        {
+            // Generated from Table D.1 of RFC3454
+            // http://www.ietf.org/rfc/rfc3454.txt
+
+            return c >= 0x0005D0 && c <= 0x0005EA ||
```

Comment:
Are you sure that these are not Unicode code points instead? You need one or two chars to build a unicode codepoint actually (a char is UTF16 in .NET but can span across two chars to get a full unicode codepoint). You should probably iterate differently on the char taking into account that

Reply:
Usually you do this by checking that the first char is `char.IsHighSurrogate(c1)`, if it is not, it is directly a codepoint otherwise you need to read the next character, then verify that it is a `char.IsLowSurrogate(c2)` and if it is, you can convert  `c1` and `c2` to `char.ConvertToUtf32(c1, c2)` to a codepoint.

Reply:
Also side note, add an internal comment to the method `IsRightToLeft` to specify that the implementation is inefficient and should probably be transformed to a binary search approach at some point.

Reply:
@xoofx I worked on your suggestion and this is what I came up with:
```
private bool StartsWithRtlCharacter(string text)
{
    for (int i = 0; i < text.Length; i++)
    {
        int c;

        if (char.IsHighSurrogate(text[i]) &&
            i < text.Length - 1 &&
            char.IsLowSurrogate(text[i + 1]))
        {
            c = char.ConvertToUtf32(text[i], text[i + 1]);
        }
        else
        {
            c = text[i];
        }

        if (CharHelper.IsRightToLeft(c))
            return true;
        else if (CharHelper.IsLeftToRight(c))
            return false;
    }

    return false;
}
```
However, this method has one big downside: `char.IsHighSurrogate`, `char.IsLowSurrogatee` and `char.ConvertToUtf32` are not supported by `portable40-net40+sl5+win8+wp8+wpa81`. I will do more research to see if there is any other way we can deal with that. There is no RTL character that needs surrogates in UTF-16 (Last character of the table is [FEFC](https://unicode-table.com/en/FEFC/), so can't we just safely ignore surrogates?

Reply:
And about performance, I stored the RTL and LTR characters in to arrays of `Range`s and wrote  a binary search for it, and you're right. Binary search seems to be faster. Here is the benchmark result:
```
BenchmarkDotNet=v0.11.1, OS=Windows 10.0.17134.228 (1803/April2018Update/Redstone4)
Intel Core i7-6700HQ CPU 2.60GHz (Skylake), 1 CPU, 8 logical and 4 physical cores
Frequency=2531247 Hz, Resolution=395.0622 ns, Timer=TSC
.NET Core SDK=2.1.400
  [Host] : .NET Core 2.1.2 (CoreCLR 4.6.26628.05, CoreFX 4.6.26629.01), 64bit RyuJIT
  Core   : .NET Core 2.1.2 (CoreCLR 4.6.26628.05, CoreFX 4.6.26629.01), 64bit RyuJIT

Job=Core  Runtime=Core

                    Method |      Mean |     Error |    StdDev | Scaled | Rank |
-------------------------- |----------:|----------:|----------:|-------:|-----:|
       ConstantsMethodText |  18.92 ns | 0.3426 ns | 0.3205 ns |   1.00 |    1 |
                           |           |           |           |        |      |
    BinarySearchMethodText |  39.23 ns | 0.8282 ns | 0.9205 ns |   1.00 |    1 |
                           |           |           |           |        |      |
    ConstantsMethodOneChar | 217.42 ns | 2.1434 ns | 1.9000 ns |   1.00 |    1 |
                           |           |           |           |        |      |
 BinarySearchMethodOneChar |  31.75 ns | 0.3695 ns | 0.3457 ns |   1.00 |    1 |
```
This is how the ranges are represented:
```
public struct Range
{
    public Range(int upper, int lower)
    {
        Upper = upper;
        Lower = lower;
    }

    public Range(int value)
    {
        Upper = Lower = value;
    }

    public int Upper { get; set; }
    public int Lower { get; set; }

    public bool Contains(int value)
    {
        return Lower >= value && Upper <= value;
    }
}
```
This is the benchmark class:
```
[CoreJob(baseline: true)]
[RPlotExporter, RankColumn]
public class Benchmarks
{
    private string _alphabet = "1234567890ABCDEFGHIJKLMNOPQRSTUVWXYZ!@#$%^&*()_+=- ~``قوەرتیئئحۆپاسدفگهژکلزخجڤبنميڕطێءكآشذىظچضةـڵأؤ";
    private char[] _characters;
    private char _character;
    private Range[] _rightToLeftCharacters;
    private Range[] _leftToRightCharacters;

    [GlobalSetup]
    public void Setup()
    {
        const int capacity = 256;

        var random = new Random();
        _characters = new char[capacity];

        for (int i = 0; i < capacity; i++)
        {
            _characters[i] = _alphabet[random.Next(_alphabet.Length)];
        }

        _character = _alphabet[random.Next(_alphabet.Length)];

        _rightToLeftCharacters = Unicode.RightToLeftCharacters;
        _leftToRightCharacters = Unicode.RightToLeftCharacters;
    }

    [Benchmark]
    public bool ConstantsMethodText()
    {
        foreach (var c in _characters)
        {
            if (IsRightToLeft(c))
                return true;
            else if (IsLeftToRight(c))
                return false;
        }

        return false;
    }

    [Benchmark]
    public bool BinarySearchMethodText()
    {
        foreach (var c in _characters)
        {
            if (ContainsBinarySearch(_rightToLeftCharacters, c))
                return true;
            else if (ContainsBinarySearch(_leftToRightCharacters, c))
                return false;
        }

        return false;
    }

    [Benchmark]
    public bool ConstantsMethodOneChar()
    {
        var isRightToLeft = IsRightToLeft(_character);
        var isLeftToRight = IsLeftToRight(_character);

        return isRightToLeft;
    }

    [Benchmark]
    public bool BinarySearchMethodOneChar()
    {
        var isRightToLeft = ContainsBinarySearch(_rightToLeftCharacters, _character);
        var isLeftToRight = ContainsBinarySearch(_leftToRightCharacters, _character);

        return isRightToLeft;
    }

    private bool ContainsBinarySearch(Range[] ranges, int c)
    {
        int low = 0;
        int high = ranges.Length - 1;

        while (low < high)
        {
            int middle = (low + high) / 2;

            if (ranges[middle].Contains(c))
            {
                return true;
            }
            if (c > ranges[middle].Upper)
            {
                low = middle;
            }
            else
            {
                high = middle;
            }
        }

        return false;
    }

    public static bool IsRightToLeft(char c)
    {
        // Generated from Table D.1 of RFC3454
        // http://www.ietf.org/rfc/rfc3454.txt

        return c >= 0x0005D0 && c <= 0x0005EA ||
              // ...other characters are omitted for brevity
              c == 0x00FB3E;
    }

    public static bool IsLeftToRight(char c)
    {
        // Generated from Table D.2 of RFC3454
        // http://www.ietf.org/rfc/rfc3454.txt
        return c >= 0x000041 && c <= 0x00005A ||
              // ...other characters are omitted for brevity
              c == 0x01D546;
    }
}
```
The benchmark project: [RtlBenchmark.zip](https://github.com/lunet-io/markdig/files/2324402/RtlBenchmark.zip)

Reply:
You can add the following code to CharHelper:

```C#
        private const char HIGH_SURROGATE_START = '\ud800';
        private const char HIGH_SURROGATE_END = '\udbff';
        private const char LOW_SURROGATE_START = '\udc00';
        private const char LOW_SURROGATE_END = '\udfff';

        public static bool IsHighSurrogate(char c)
        {
            return ((c >= HIGH_SURROGATE_START) && (c <= HIGH_SURROGATE_END));
        }

        public static bool IsLowSurrogate(char c)
        {
            return ((c >= LOW_SURROGATE_START) && (c <= LOW_SURROGATE_END));
        }
```

Reply:
Do you still need to support portable profiles such as `ortable40-net40+sl5+win8+wp8+wpa81`? I thought that with .net standard those would be gone (and win8 is a dead OS anyway).

Reply:
@xoofx  and for `ConvertToUtf32`, can we use .net [core's code](https://github.com/dotnet/corefx/blob/dc54c458e905d282c0c453ce0b930e363fe84069/src/Common/src/CoreLib/System/Char.cs#L1007)?

```
public static int ConvertToUtf32(char highSurrogate, char lowSurrogate)
{
    if (!IsHighSurrogate(highSurrogate))
    {
        throw new ArgumentOutOfRangeException(nameof(highSurrogate), SR.ArgumentOutOfRange_InvalidHighSurrogate);
    }
    if (!IsLowSurrogate(lowSurrogate))
    {
        throw new ArgumentOutOfRangeException(nameof(lowSurrogate), SR.ArgumentOutOfRange_InvalidLowSurrogate);
    }
    return (((highSurrogate - CharUnicodeInfo.HIGH_SURROGATE_START) * 0x400) + (lowSurrogate - CharUnicodeInfo.LOW_SURROGATE_START) + UNICODE_PLANE01_START);
}
```

Reply:
yes (we will probably have to copy the MIT license in CharHelper at least)

Reply:
> Do you still need to support portable profiles such as portable40-net40+sl5+win8+wp8+wpa81? I thought that with .net standard those would be gone (and win8 is a dead OS anyway).

Well, I would not, but I'm worried that it may break a few apps out there, so I would prefer that we keep this compatibility for a period and I will probably kill it at some point